polymorphic.h

```diff
@@ -31,45 +31,64 @@ namespace xyz {
 namespace detail {
 template <class T, class A>
 struct control_block {
-  T* p_;
-  virtual ~control_block() = default;
-  virtual void destroy(A& alloc) = 0;
-  virtual control_block<T, A>* clone(A& alloc) = 0;
+  typedef void(destroy_fn)(control_block*, A& alloc);
+  typedef control_block<T, A>*(clone_fn)(const control_block*, A& alloc);
+
+  struct vtable {
+    destroy_fn* const destroy;
+    clone_fn* const clone;
+  } local_vtable_;
+
+  control_block(destroy_fn* const destroy, clone_fn* const clone)
+      : local_vtable_{.destroy = destroy, .clone = clone} {}
+
+  void destroy(A& alloc) { local_vtable_.destroy(this, alloc); }
+
+  control_block* clone(A& alloc) const {
+    return local_vtable_.clone(this, alloc);
+  }
+};
+
+template <class T, class A>
+struct fake_control_block : public control_block<T, A> {
+  fake_control_block() = delete;
+  std::byte data_;
 };
 
 template <class T, class U, class A>
 class direct_control_block : public control_block<T, A> {
   U u_;
```

Comment:
I don't think this will work, if I understand how you want it to work: If U has an alignment higher than sizeof(void*) the offset to fake_control_block::data_ will not be the same as the offset to u_ here.